lib/Core/Inference/FrameBuilder/AbstractInstanceOfWalker.php

```diff
@@ -67,24 +83,73 @@ protected function variableFromBinaryExpression(BinaryExpression $node)
 
         $type = (string) $rightOperand->getResolvedName();
 
-        $context = $this->createSymbolContext($variable);
+        $context = $this->createSymbolContext($variable, $frame);
         $context = $context->withType(Type::fromString($type));
         $variable = WorseVariable::fromSymbolContext($context);
 
         return $variable;
     }
 
-    protected function createSymbolContext(Variable $leftOperand)
+    protected function createSymbolContext(Expression $leftOperand, Frame $frame): SymbolContext
     {
-        $context = $this->symbolFactory()->context(
-            $leftOperand->getName(),
+        assert($leftOperand instanceof Variable || $leftOperand instanceof MemberAccessExpression);
+
+        if ($leftOperand instanceof MemberAccessExpression) {
+            return $this->createPropertySymbolContext($leftOperand, $frame);
+        }
+
+        return $this->createVariableSymbolContext($leftOperand);
+    }
+
+    protected function createPropertySymbolContext(
+        MemberAccessExpression $leftOperand,
+        Frame $frame
+    ): SymbolContext {
+        assert($leftOperand->dereferencableExpression instanceof Variable);
+
+        $classVariableName = $leftOperand->dereferencableExpression->getName();
+
+        try {
+            $classType = $frame->locals()->byName($classVariableName)->first()
+                ->symbolContext()->types()->best()
+            ;
+        } catch (\RuntimeException $exception) {
+            // TODO log the fact that the variable was not recognize ?
+            $classType = Type::unknown();
```

Comment:
If it's something that could be classed as a coding error, you _could_ add it as an "issue" in the symbol context (`SymbolContext#withIssue`) - but it won't get used for anything currently - however that might change (the current LS diagnostics are very inefficient, and using these diagnostics may be far more efficient, but will require some rework and experimentation).

Reply:
Done, I let you resolve if what I did match what you had in mind.